components/bin/build

```diff
@@ -0,0 +1,318 @@
+    while (PACKAGE.length) {
+        const dir = path.dirname(PACKAGE[0]).split(path.sep)[0];
+        packages.push(processPackage(lines, '    ', dir));
+    }
+    lines.push('    ' + packages.join(',\n    '));
+    lines.push('}});');
+    fs.writeFileSync(path.join(LIB, COMPONENT + '.js'), lines.join('\n') + '\n');
+}
+
+/**
+ * Sort file paths alphabetically
+ */
+function sortDir(a, b) {
+    const A = a.replace(/\//g, '|');          // Replace directory separator by one that is after the alphnumerics
+    const B = b.replace(/\//g, '|');
+    return (A === B ? 0 : A < B ? -1 : 1);
+ }
+
+/**
+ * Recursively process packages, collecting them into groups by subdirectory, properly indented
+ *
+ * @param {string[]} lines   The array where lines of code defining the packages are to be stored
+ * @param {string} space     The current level of indentation
+ * @param {string} dir       The subdirectory currently being processed
+ * @return {string}          The string to use to load all the objects from the given directory
+ */
+function processPackage(lines, space, dir) {
+    const packages = [];
+    //
+    //  Loop through the lines that are in the current directory
+    //
+    while (PACKAGE.length && (PACKAGE[0].substr(0, dir.length) === dir || dir === '.')) {
+        //
+        // If the current package is in this directory (not a subdirectory)
+        //   Get the location of transpiled mathjax file
+        //   Get the name to use for the data from that file
+        //   Create an entry for the file in the MathJax global that loads the reuired component
+        // Otherwise (its in a subdirectory)
+        //   Get the subdirectory name
+        //   Process the subdirectory using an additional indentation
+        //
+        if (path.dirname(PACKAGE[0]) === dir) {
+            const file = PACKAGE.shift();
+            const name = path.basename(file);
+            const relativefile = path.join('..', MATHJAX3, dir, name).replace(/\.ts$/, '.js');
+            const component = `require('${relativefile}')`;
+            let property = name.replace(/\.ts$/, '');
+            if (property !== name && exists(path.resolve(MATHJAX3, file.replace(/\.ts$/, '')))) property += '_ts';
+            if (property.match(/[^a-zA-Z0-9_]/)) {
+                property = `"${property}"`;
+            }
+            packages.push(`${property}: ${component}`);
+        } else {
+            let subdir = path.dirname(PACKAGE[0]);
+            while (path.dirname(subdir) !== dir && subdir !== '.') {
+                subdir = path.dirname(subdir);
+            }
+            packages.push(processPackage(lines, space + (dir === '.' ? '' : '    '), subdir));
+        }
+    }
+    //
+    // Create the string defining the object that loads all the needed files into the proper places
+    //
+    if (dir === '.') return packages.join(',\n    ');
+    return path.basename(dir) + ': {\n    ' + space + packages.join(',\n    '+space) + '\n' + space + '}';
```

Comment:
I note there is a mix of string composition syntax here. Maybe we should just switch to the new one everywhere, starting with this file.

Reply:
Otherwise there missing spaces around `+`

Reply:
I think this is one of those things that needs to be done as a global update.